query-engine/core/src/metrics/registry.rs

```diff
@@ -0,0 +1,307 @@
+        let key = Key::from_name(name);
+        let gauges = self.inner.register.get_gauge_handles();
+        let gauge = gauges.get(&key)?;
+        let val = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
+        Some(val)
+    }
+
+    pub fn histogram_values(&self, name: &'static str) -> Option<HistogramUtil> {
+        let mut histogram = HistogramUtil::new(&HISTOGRAM_BOUNDS)?;
+        let key = Key::from_name(name);
+        let histograms = self.inner.register.get_histogram_handles();
+        let samples = histograms.get(&key)?;
+
+        samples.get_inner().data_with(|s| {
+            histogram.record_many(s);
+        });
+        Some(histogram)
+    }
+
+    pub fn get_descriptions(&self) -> HashMap<String, String> {
+        let descriptions = self.inner.descriptions.read();
+        descriptions.clone()
+    }
+
+    fn get_snapshot(&self, global_labels: IndexMap<String, String>) -> Snapshot {
+        let counter_handles = self.inner.register.get_counter_handles();
+        let gauge_handles = self.inner.register.get_gauge_handles();
+        let histogram_handles = self.inner.register.get_histogram_handles();
+        let descriptions = self.get_descriptions();
+
+        let mut counters: Vec<Metric> = counter_handles
+            .into_iter()
+            .map(|(key, counter)| {
+                let key_name = key.name();
+                let value = counter.get_inner().load(Ordering::Acquire);
+                let description = descriptions.get(key_name).cloned().unwrap_or_default();
+                Metric::new(key, description, MetricValue::Counter(value), global_labels.clone())
+            })
+            .collect();
+
+        let mut gauges: Vec<Metric> = gauge_handles
+            .into_iter()
+            .map(|(key, gauge)| {
+                let key_name = key.name();
+                let description = descriptions.get(key_name).cloned().unwrap_or_default();
+                let value = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
+                Metric::new(key, description, MetricValue::Gauge(value), global_labels.clone())
+            })
+            .collect();
+
+        let mut histograms: Vec<Metric> = histogram_handles
+            .into_iter()
+            .map(|(key, samples)| {
+                let mut histogram = HistogramUtil::new(&HISTOGRAM_BOUNDS).unwrap();
+                samples.get_inner().data_with(|s| {
+                    histogram.record_many(s);
+                });
+
+                let key_name = key.name();
+                let description = descriptions.get(key_name).cloned().unwrap_or_default();
+                Metric::new(
+                    key,
+                    description,
+                    MetricValue::Histogram(histogram.into()),
+                    global_labels.clone(),
+                )
+            })
+            .collect();
+
+        // Sort them so that they are in ordered by key name
```

Comment:
Why is this required?

Reply:
It makes sure that the metrics are always in the same order in the array. It is most useful for testing but I think probably a good thing for our users as well.